### PR TITLE
Allow `del adata.X`, `adata.X = None`

### PR DESCRIPTION
**Pull request by rcannood**
Feature proposal for #464.

Need to test:
- [x] Getter (`test_base.py > test_x_is_none()`)
- [x] Setter (`test_base.py > test_x_is_none()`)
- [x] Deleter (`test_base.py > test_x_is_none()`)
- [x] Initializer (`test_base.py > test_x_is_none()`)
- [ ] Views
    * [ ] Check that .X = None doesn't change other view behavior
    * [ ] a[idx].X doesn't do anything weird
    * [ ] copying views still works
- [ ] _subset_inplace
- [ ] Copying
- [ ] Backed mode
- [ ] Concatenation
- [x] Transpose
- [ ] Methods which default to values from X, e.g. obs_vector, to_df.
- [ ] IO (seems to work, needs testing)
    * [ ] Maybe specialized writers like to_csvs will need some attention here?
- [ ] Raw (pretty similar set of things to be checked)
    * [ ] Initializer
    * [ ] Backed
    * [ ] Copies
    * [ ] Concatenation
    * [ ] IO

### CODE REVIEW

**Comment by ivirshup** [comment]
I don't think we currently allow `X` to be `None` (#331).

I have come across a case recently where this would be useful, though I think it would take some doing to make the entire codebase work again if we allowed this. `X` being an array with `shape` is pretty baked into the data model at this point.

Maybe this could be opened as a feature request issue, and you could pitch your use case?

**Comment by rcannood** [comment]
Thanks for your input.

`X` being `None` is already part of the unit tests: [test_base.py#L46-L47](https://github.com/theislab/anndata/blob/cf915f51f6d53ee621e9c200cd0aa56fc82d8661/anndata/tests/test_base.py#L46-L47). Also, if `X` would not be allowed to be `None`, why is there a check in the code I changed? ;)

A possible use case I have is when working with read-only h5ad files: if your input is immutable, it may be interesting to exclude X from the output file to reduce the file size.

**Comment by ivirshup** [comment]
At one point it was allowed, but I can say for certain that a lot of code has been written since then that does not expect `X` to be `None`.

For example:

```python
a = ad.AnnData(X=None, shape=(100, 200))

a.copy()
```

<details>
<summary> Traceback </summary>


```pytb
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-12-02bb7baf10a6> in <module>
----> 1 a.copy()

~/github/anndata/anndata/_core/anndata.py in copy(self, filename)
   1428                 X = _subset(self._adata_ref.X, (self._oidx, self._vidx)).copy()
   1429             else:
-> 1430                 X = self.X.copy()
   1431             # TODO: Figure out what case this is:
   1432             if X is not None:

AttributeError: 'NoneType' object has no attribute 'copy'
```

</details>

At one point I think I saw `_X` being `None` as an indicator for if the object was in backed mode.

If this is something we were to support more, it would need a fair amount of testing.

**Comment by rcannood** [comment]
Alright :) I hereby turn this PR into a feature request. I'd like `X` to be optional. I mean, it [already is](https://github.com/theislab/anndata/blob/cf915f51f6d53ee621e9c200cd0aa56fc82d8661/anndata/_core/anndata.py#L283), but apparently more thorough testing is needed to verify that it works properly. 

What is the opinion of the major anndata developers? :+1: or :-1:?

If we agree `X` should be able to be `None`, we can use this PR to add tests and fix the code where necessary.

**Comment by ivirshup** [comment]
Oops, I linked the wrong issue before. #330 is where this has come up before. Also I'm going to open this discussion as a new issue so it's more visible.

File: anndata/tests/test_base.py [hunk 1]
```diff
@@ -581,3 +581,28 @@ def assert_eq_not_id(a, b):
...
+    shape = (3, 5)
+    adata = AnnData(None, uns=dict(test=np.array((3, 3))), shape=shape)
+    assert adata.X is None
+    assert adata.shape == shape
+
+    # test transpose
+    adataT = adata.transpose()
+    assert_equal(adataT.shape, (5, 3))
+    assert_equal(adataT.obsp.keys(), adata.varp.keys())
+    assert_equal(adataT.T, adata)
```
Comment (ivirshup) [inline anndata/tests/test_base.py:608]: I think this might actually be a good time to expand the tests for transpose. Mind if we do this in a separate PR?

Comment (ivirshup) [reply anndata/tests/test_base.py:608]: #468 should set the basis for this

**Comment by rcannood** [comment]
> Plus it's much easier to review and merge small PRs. Let's keep that PR's scope to just fixing up the .X property.

@ivirshup Alright, shall we merge this PR in order to resolve the first box in #467?

File: anndata/_core/anndata.py [hunk 2]
```diff
@@ -1268,7 +1273,17 @@ def transpose(self) -> "AnnData":
...
+            else:
+                return m.T
+
+        if X is not None:
+            dtype = X.dtype
+        else:
+            dtype = "float32"
 
         return AnnData(
             t_csr(X),
```
Comment (ivirshup) [inline anndata/_core/anndata.py:1289]: Instead adding all that stuff to `t_csr`, could this be: `t_csr(X) if X is not None else None`

My thinking here is that I'm not sure if we should keep the behavior of `t_csr`, so I don't want to put more logic in it.

Comment (rcannood) [reply anndata/_core/anndata.py:1289]: Fair enough!

File: anndata/_core/anndata.py [hunk 3]
```diff
@@ -683,6 +684,10 @@ def X(self, value: Optional[Union[np.ndarray, sparse.spmatrix]]):
                 f"need to be {self.shape}."
             )
 
+    @X.deleter
+    def X(self):
+        self.X = None
```
Comment (ivirshup) [inline anndata/_core/anndata.py:689]: With other attributes, deleting them when an object is a view causes it to become an "actual" object. This happens for any modification. For example:

```python
In [1]: import scanpy as sc
pbmc
In [2]: pbmc = sc.datasets.pbmc3k_processed()

In [3]: v = pbmc[:, :50]

In [4]: v
Out[4]: 
View of AnnData object with n_obs × n_vars = 2638 × 50
    obs: 'n_genes', 'percent_mito', 'n_counts', 'louvain'
    var: 'n_cells'
    uns: 'draw_graph', 'louvain', 'louvain_colors', 'neighbors', 'pca', 'rank_genes_groups'
    obsm: 'X_pca', 'X_tsne', 'X_umap', 'X_draw_graph_fr'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'

In [5]: del v.obsm

In [6]: v
Out[6]: 
AnnData object with n_obs × n_vars = 2638 × 50
    obs: 'n_genes', 'percent_mito', 'n_counts', 'louvain'
    var: 'n_cells'
    uns: 'draw_graph', 'louvain', 'louvain_colors', 'neighbors', 'pca', 'rank_genes_groups'
    varm: 'PCs'
    obsp: 'distances', 'connectivities'
```

That should happen here too. Could you add a test similar to `test_attr_deletion` which checks for this?

I think it's fair to say that `adata.X = None` should be equivalent to `del adata.X`, so you could also make sure that works when the `AnnData` is a view? It looks like it's set to throw an error right now.

Comment (ivirshup) [reply anndata/_core/anndata.py:689]: Oh, the lines where I should have commented this are 628-631, where:

```python
        if value is None:
             if self.is_view:
                 raise ValueError(
                    "Copy the view before setting the data matrix to `None`."
                )
```

Comment (rcannood) [reply anndata/_core/anndata.py:687]: See 72618c3 -- is this what you meant?

File: anndata/_core/anndata.py [hunk 4]
```diff
@@ -627,9 +627,7 @@ def X(self) -> Optional[Union[np.ndarray, sparse.spmatrix, ArrayView]]:
     def X(self, value: Optional[Union[np.ndarray, sparse.spmatrix]]):
         if value is None:
             if self.is_view:
-                raise ValueError(
-                    "Copy the view before setting the data matrix to `None`."
-                )
+                self._init_as_actual(self.copy())
```
Comment (ivirshup) [inline anndata/_core/anndata.py:630]: I think this should error if the view is of a backed object, since I'm not sure we have the logic to deal with that at the moment.

File: anndata/tests/test_base.py [hunk 5]
```diff
@@ -118,6 +118,15 @@ def test_attr_deletion():
     assert_equal(full, empty, exact=True)
 
 
+def test_not_view_after_deletion():
+    full = gen_adata((30, 30))
+    for attr in ["X", "obs", "var", "obsm", "varm", "obsp", "varp", "layers", "uns"]:
+        view = full[:, :20]
+        assert view.is_view
+        delattr(view, attr)
+        assert not view.is_view
```
Comment (ivirshup) [inline anndata/tests/test_base.py:127]: Could this instead be a modification of `test_view_delattr` in `test_views.py`?

Comment (ivirshup) [reply anndata/tests/test_base.py:127]: In general, I think `test_base.py` covers too much API surface, and has too much in it. When possible, try to add tests in more specific files. Should make it easier to find tests in the future.

File: anndata/_core/anndata.py (hunk 4 above)
Comment (ivirshup) [reply anndata/_core/anndata.py:630]: I think these checks need to be flipped, so the lower error get's thrown, not an error from trying to copy.

At the moment:

```python
from anndata.tests.helpers import gen_adata
import anndata as ad

gen_adata((10, 20)).write_h5ad("tmp.h5ad")
b = ad.read_h5ad("tmp.h5ad", backed="r")
v = b[::2, ::2]
del v.X
```

```pytb
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-11-6b92c05b346f> in <module>
----> 1 del v.X

~/github/anndata/anndata/_core/anndata.py in X(self)
    685     @X.deleter
    686     def X(self):
--> 687         self.X = None
    688 
    689     @property

~/github/anndata/anndata/_core/anndata.py in X(self, value)
    628         if value is None:
    629             if self.is_view:
--> 630                 self._init_as_actual(self.copy())
    631             if self.isbacked:
    632                 raise ValueError("Not implemented.")

~/github/anndata/anndata/_core/anndata.py in copy(self, filename)
   1460 
   1461             if filename is None:
-> 1462                 raise ValueError(
   1463                     "To copy an AnnData object in backed mode, "
   1464                     "pass a filename: `.copy(filename='myfilename.h5ad')`."

ValueError: To copy an AnnData object in backed mode, pass a filename: `.copy(filename='myfilename.h5ad')`.
```

Also the error should be more descriptive, and probably be an actual `NotImplementedError`.

Maybe:

```python
raise NotImplementedError("Cannot currently remove data matrix from backed object.")
```

Comment (rcannood) [reply anndata/_core/anndata.py:632]: Done! :)

File: anndata/_core/anndata.py [hunk 6]
```diff
@@ -625,20 +625,21 @@ def X(self) -> Optional[Union[np.ndarray, sparse.spmatrix, ArrayView]]:
...
     def X(self, value: Optional[Union[np.ndarray, sparse.spmatrix]]):
-        if not isinstance(value, StorageType.classes()) and not np.isscalar(value):
-            if hasattr(value, "to_numpy") and hasattr(value, "dtypes"):
-                value = ensure_df_homogeneous(value, "X")
-            else:  # TODO: asarray? asanyarray?
-                value = np.array(value)
         if value is None:
-            if self.is_view:
+            if self.isbacked:
                 raise ValueError(
```
Comment (ivirshup) [inline anndata/_core/anndata.py:630]:
```suggestion
                raise NotImplementedError(
```

**Comment by ivirshup** [comment]
Great, thanks for implementing this and making the suggestion!
